File: euclidean_primer/euclidean_primer__rsa_basic_principle.py
# 実行時間 10ms
def extGcd(a, b):
    if b != 0:
        c, y, x = extGcd(b, a % b)

        y -= a // b * x
        return c, x, y
    return a, 1, 0

p, q, e, m = list(map(int, input().split()))
n = p * q
c = (p - 1) * (q - 1)
x = extGcd(e, c)[1]
d = (c + x) % c
e = pow(m, e, n)
f = pow(e, d, n)

print(d)
print(e)
print(f)

File: euclidean_primer/euclidean_primer__rsa_basic_principle.cs
using static System.Console;

internal class Program
{
    static void Main()
    {
        long[] pqeM = System.Array.ConvertAll( ReadLine().Split(), long.Parse );
        long n = pqeM[0] * pqeM[1];
        long c = ( pqeM[0] - 1 ) * ( pqeM[1] - 1 );
        long x = 1;
        long y = 1;

        ExtGcd( pqeM[2], c, ref x, ref y );

        long d = ( c + x ) % c;
        ulong e = ModPow( pqeM[3], pqeM[2], n );
        ulong f = ModPow( (long)e, d, n );

        WriteLine( d );
        WriteLine( e );
        WriteLine( f );
        return;
    }

    static long ExtGcd( long a, long b, ref long x, ref long y )
    {
        long c = a;

        if( b != 0 ) {
            c = ExtGcd( b, a % b, ref y, ref x );
            y -= a / b * x;
        } else {
            x = 1;
            y = 0;
        }
        return c;
    }

    static ulong ModPow( long a, long b, long m )
    {
        ulong la = (ulong)a;
        ulong lb = (ulong)b;
        ulong lm = (ulong)m;
        ulong o = 1;

        while( 0 < lb ) {
            if( ( lb & 1 ) == 1 ) {
                o = o * la % lm;
            }
            la = la * la % lm;
            lb >>= 1;
        }
        return o;
    }
}
